source-postgres-batch/main.go

```diff
@@ -105,46 +106,55 @@ func connectPostgres(ctx context.Context, cfg *Config) (*sql.DB, error) {
 	return db, nil
 }
 
-const tableQueryTemplateTemplate = `{{if .IsFirstQuery -}}
-  SELECT xmin AS txid, * FROM %[1]s ORDER BY xmin::text::bigint;
+// A discussion on the use of XIDs as query cursors:
+//
+// XID values are 32-bit unsigned integers with implicit wraparound on overflow and underflow.
+// The lowest 'normal' XID value is 3. The values 0-2 are reserved as sentinels, with 2 being
+// the "Frozen XID" value.
+//
+// While Postgres has a comparison function `TransactionIdPrecedes()` internally, this is not
+// exposed in the form of a comparison predicate or ordering method. So we have to cast XIDs
+// to integers (by way of text) and then reimplement the desired ordering behavior ourselves.
+//
+// Given a particular "cursor" value obtained from a previous polling query execution, we can
+// begin by assuming that there are fewer than 2^31 "live" XIDs greater than that cursor. The
+// issue is that with XID wraparound some of these values may be numerically smaller. Given a
+// wrapping uint32 wrapping subtraction operator the expression `xmin - cursor` would compute
+// an orderable count of "how far past the previous cursor" a given row is.
+//
+// We can emulate wrapping uint32 subtraction using PostgreSQL int64 values by simply writing
+// the expression `((x::bigint - y::bigint)<<32)>>32`.
```

Comment:
I don't really understand how this is a wrapping subtraction, can you give a numerical example of a wrapped `xid` value and a `cursor` value and show how this calculation unfolds in practice please?

Reply:
Suppose we have a table with four rows:
* Row A has `xmin = 0xFFFFC000`
* Row B has `xmin = 0xFFFFF000`. This is our cursor value, in this example.
* Row C has `xmin = 0xFFFFFFFF`
* Row D has `xmin = 0x00001000`

The correct comparison operation on XIDs is the u32 comparison-via-subtraction-and-sign-check function `x - y > 0`. In normal non-wrapping cases it should be obvious that this works because it's equivalent to `x > y`, and in wrapping cases we get the following results:

- Row A `0xFFFFC000 - 0xFFFFF000 = -0x3000`
- Row B `0xFFFFF000 - 0xFFFFF000 = 0`
- Row C `0xFFFFFFFF - 0xFFFFF000 = 0xFFF`
- Row D `0x00001000 - 0xFFFFF000 = 0x2000`

The `(x<<32)>>32` operation takes the bottom 32 bits of an int64 quantity, shifts them up to the top of the int64 so that the 32-bit MSB becomes the int64's MSB, and then shifts them back down to the normal position. This is a fairly common idiom for "sign-extend an N-bit quantity into an M-bit one (where M > N)".

Reply:
Working out the math on that final one in more detail, we have something like:

```
(((0x00001000::bigint - 0xFFFFF000::bigint)<<32)>>32)::int
(((0x0000000000001000 - 0x00000000FFFFF000)<<32)>>32)::int
((0xFFFFFFFF00002000<<32)>>32)::int
(0x0000200000000000>>32)::int
0x0000000000002000::int
0x00002000
```

But the final cast back to `::int` is irrelevant and unnecessary for our purposes (we just want to filter for positive results and order by the result value) so we skip it.